### ISSUE TITLE
add_thin_conductor PEC sheet does not act as a resonant reflector (pec_mask set, inv_eps not zeroed)

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#369]
## Symptom

A `add_thin_conductor` copper sheet (σ≥1e6 → routed to `pec_mask`) does **not** reflect/resonate like an equivalent PEC `Box` at the same location. In a patch-cavity setup, the PEC-Box patch **rings** while the thin-conductor patch **decays** (~700× smaller sustained field), so Harminv finds no mode.

## Evidence (air-patch cavity, dx=1.5mm, 400 steps, identical geometry except the patch)

| patch model | last-50-step max\|Ez\| | behavior |
|---|---|---|
| PEC `Box` | 5.7e5 | rings (sustained) |
| `add_thin_conductor` (copper) | 8.3e2 (=1/700 of peak) | decays |

`pec_mask` is **identical** for both (verified via `_assemble_materials`: same 1178 cells, same z-distribution, patch cell `pec_mask=True` in both).

## Root cause

`apply_thin_conductor` for the PEC branch (`rfx/materials/thin_conductor.py:95-101`) returns `materials` **unchanged** and only updates `pec_mask`. The uniform runner's **primary** PEC freeze is `inv_eps=0` from the geometry (`rfx/runners/uniform.py:89`); `pec_mask` is documented "defense-in-depth" on top of that (`:105-112`). A PEC `Box` sets both (inv_eps=0 + pec_mask) → strong reflector. The thin conductor sets **only** `pec_mask` (inv_eps stays ≈1, so E keeps updating normally), and the pec_mask defense-in-depth alone is insufficient to sustain a resonance → the field leaks away.

## Impact

`add_thin_conductor` cannot currently be used to model a radiator/patch conductor sheet (e.g. the planar-radiator thin-sheet accuracy direction, #330). Its PEC-sheet reflection was never validated — `tests/test_thin_conductor.py` covers σ_eff / mask selectivity / API integration, but no test checks that the sheet actually reflects or resonates.

## Fix direction

For the PEC branch, also zero `inv_eps` (the primary freeze) at the masked cells, not just `pec_mask` — OR make the runner's pec_mask enforcement a true tangential-E freeze sufficient on its own. Add a resonance/reflection regression test (a thin-conductor patch cavity must ring like the PEC-Box equivalent).

_Found during the RA-2.2 thin-sheet investigation (2026-07-14); full evidence in the local research note 20260713_research_agenda_principle_based.md "RA-2.2 option-1 (thin-sheet) — BLOCKED"._

**Comment by bk-squared** [comment]
## Correction + sharper reproduction (2026-07-14)

The original title/framing ("inv_eps=0") is **wrong** — corrected findings below. Two adversarial trace passes (one via a teammate static-trace) refuted both the inv_eps hypothesis AND a cell-selection hypothesis.

### Sharpest reproduction (control)
Using the **identical zero-extent `Box` object** for both APIs, same cavity, 300 steps:
- `sim.add(box, material="pec")` (geometry-PEC) → cavity **RINGS** (last-50 max\|Ez\| ≈ 7e5)
- `sim.add_thin_conductor(box, ...)` (thin) → cavity **DECAYS** (last-50 ≈ 1.5e3)
- `max\|Ez_box − Ez_thin\|` over the run = **1.36e6**. Determinism verified clean (box-vs-box diff = 0.00).

So a PEC-Box and an `add_thin_conductor` with the **same shape** behave differently — the sheet does not sustain a cavity resonance.

### Bit-identical between the two APIs (verified via `_assemble_materials`, diff = 0 on every array)
`eps_r`, `sigma`, `mu_r`, **`pec_mask`** all bit-identical. `pec_shapes` = `[ground Box, patch Box]` in both.

### Ruled out (with evidence)
- **inv_eps=0** — geometry PEC does NOT zero inv_eps either (keeps vacuum eps, `_compile.py:157-159`); the inv_eps=0 mechanism only exists in the opt-in `subpixel_smoothing="kottke_pec"` path. So this was never the mechanism.
- **cell-selection inconsistency** (`csg.py:104-108`, thin-branch `argmin` vs volume-branch half-open) — a **real separate bug** (a thin conductor and a matching-thickness volume box at the same nominal position can select different cells), but NOT this symptom's cause: here `pec_mask` is bit-identical.
- **materials / pec_mask lost in the run** — no; `_execute.py:2462` assembles once (incl. thin via `_compile.py:230`) and passes it through.
- **H-freeze** (`apply_pec_h_mask`) — gated by `ctx.use_aniso_inv` (kottke only, off here).
- **`_thin_conductors` / `_geometry` read directly in the run path** — grep-empty except the subpixel/kottke blocks (`uniform.py:97/126`, off here).

### The paradox (open)
Identical assembled materials + `pec_mask`, standard run path does not read `_geometry`/`_thin_conductors`, deterministic runner — **yet geometry-PEC rings and thin-conductor decays**. So either the run branches on which list the patch came from, or the run's ACTUAL `pec_mask` at the compiled step differs from what `_assemble_materials` returns. Runtime instrumentation is blocked by jit closure capture; needs static tracing or `jax.debug.print` inside the compiled step.

**Status:** real, precisely characterized, mechanism not yet pinned. Under active trace. The `csg.py` two-rule inconsistency should also be fixed on its own merits (separate).

**Comment by bk-squared** [comment]
## Trigger isolated: dz_profile / domain-z=0 path — NOT csg rasterization (2026-07-14, update 2)

A bisection (add elements to a minimal cavity, discriminate by late-time energy) **narrows the bug decisively** and **refutes the csg two-rule root cause for this symptom**.

### Decisive isolation
| setup | box last-50 | thin last-50 | ratio | verdict |
|---|---|---|---|---|
| **uniform 3D** (real z-domain, `domain=(L,L,L)`, plain `dx`) | 1.27e-5 | 1.27e-5 | **1.000** | **box ≡ thin — thin mechanism CORRECT** |
| **dz_profile** (`domain=(L,L,0)` + `dz_profile`) generic cube | 2.77e3 | 1.02e3 | 0.369 | thin diverges (partial) |
| dz_profile, t1 air-patch (tuned to resonate) | 7.9e5 | 1.6e4 | 0.02 | box rings Q=15 @ 3.82 GHz, **thin NO MODE** (harminv) |

`add_thin_conductor` and a geometry-PEC `Box` are **bit-identical in a real uniform 3D domain**. They diverge **only in the `dz_profile` / `domain z=0` ("2.5D") configuration** — which is exactly the config planar-radiator work needs (coarse-dz air gap).

### csg two-rule root cause is REFUTED for this symptom
`build_z_uniform` returns a **uniform** dz (1.5 mm, all cells). At the patch z, all three rasterization rules agree:
- thin-branch `argmin` → cell k10
- proposed half-open fix → cell k10
- volume-branch (box) → cell k10

So `pec_mask` is bit-identical and the csg.py:104/108 inconsistency **does not fire here** (it needs a nonuniform grid or a thin/volume thickness mismatch to select different cells). The csg inconsistency may still be a **latent bug for genuinely nonuniform grids** — worth a separate targeted test/fix — but it is **not** the mechanism of this reflection failure.

### Still-open paradox (now scoped)
In the `dz_profile` path, `eps_r`/`sigma`/`mu_r` (→ `inv_eps`) and `pec_mask` are all bit-identical between the two APIs, yet dynamics diverge. So the divergence is in the **`dz_profile` grid-assembly or run path**, reading `_geometry` vs `_thin_conductors` (or building the z-grid) differently — not in cell selection, not in enforcement, not in assembled materials. Runtime instrumentation is blocked by jit closure capture; needs a `dz_profile`-specific static trace or `jax.debug.print` inside the step.

**Status:** real, precisely scoped to the `dz_profile`/`domain z=0` path, mechanism not yet pinned. **No fix committed** — the csg fix is a no-op for this symptom, and I will not land a change that doesn't demonstrably close it. Next step: focused `dz_profile` assembly/run-path trace.

**Comment by bk-squared** [comment]
## ROOT CAUSE FOUND + FIXED → PR #370 (2026-07-14, resolution)

The mechanism is **not** in cell selection, enforcement, or assembled state — it is that the **non-uniform (`dz_profile`) runner never rasterized thin conductors at all**.

`rfx/runners/nonuniform.py::assemble_materials_nu` passed only `sim._geometry` to `rasterize_geometry`, with an explicit shortcut:
```python
# Thin conductors require mask(grid) which needs uniform Grid.
# Skip on NU path for now — the shape won't resolve.
```
So every `add_thin_conductor()` on a `dz_profile` grid was **silently dropped** → no `pec_mask` → the sheet neither reflected nor resonated. Geometry PEC (in `sim._geometry`) *was* rasterized → rang.

This also explains why my earlier dumps looked bit-identical: they used the **uniform** assembly (`_compile.py::_assemble_materials`, which *does* apply thin conductors), while the actual `dz_profile` run used `assemble_materials_nu`, which did not. Bisection confirmed it: `add_thin_conductor` ≡ box-PEC in a real uniform-3D domain (ratio 1.000), diverges only under `dz_profile`.

The `mask_on_coords` shortcut premise was wrong — it resolves fine on non-uniform axes (geometry PEC already uses it there). Fix rasterizes PEC thin conductors into `pec_mask` the same way; lossy sheets now warn instead of dropping.

**Verified:** thin rings identically to box on the NU path (ratio 1.000, was ≤0.37); AD preserved (63 NU grad/forward tests green — boolean-mask OR adds no float to the gradient path); NU + thin_conductor suites green; two new regression tests. Closing on merge of #370.
